pkg/operator/ceph/object/config.go

```diff
@@ -126,16 +169,119 @@ func (c *clusterConfig) setFlagsMonConfigStore(rgwConfig *rgwConfig) error {
 	configOptions["rgw_zone"] = rgwConfig.Zone
 	configOptions["rgw_zonegroup"] = rgwConfig.ZoneGroup
 
+	configOptions, err := configureKeystoneAuthentication(rgwConfig, configOptions)
+	if err != nil {
+		return err
+	}
+
+	s3disabled := false
+	if s3 := rgwConfig.Protocols.S3; s3 != nil {
+		if s3.Enabled != nil && !*s3.Enabled {
+			s3disabled = true
+		}
+
+		if s3.AuthUseKeystone != nil {
+			configOptions["rgw_s3_auth_use_keystone"] = fmt.Sprintf("%t", *s3.AuthUseKeystone)
+		}
+
+	}
+
+	if swift := rgwConfig.Protocols.Swift; swift != nil {
+
+		if swift.AccountInUrl != nil {
+			configOptions["rgw_swift_account_in_url"] = fmt.Sprintf("%t", *swift.AccountInUrl)
+		}
+
+		if swift.UrlPrefix != nil {
+			configOptions["rgw_swift_url_prefix"] = *swift.UrlPrefix
+
+			if configOptions["rgw_swift_url_prefix"] == "/" {
+				logger.Warning("Forcefully disabled S3 as the swift prefix is given as a slash /. Ignoring any S3 options (including Enabled=true)!")
+				// this will later on disable the s3 api using the rgw_enable_apis setting
+				s3disabled = true
+			}
+
+		}
+		if swift.VersioningEnabled != nil {
+			configOptions["rgw_swift_versioning_enabled"] = fmt.Sprintf("%t", *swift.VersioningEnabled)
+		}
+
+	}
+
+	if s3disabled {
+		// XXX: how to handle enabled APIs? We only configure s3 and
+		// swift in the resource, `admin` is required for the operator to
+		// work, `swift_auth` is required to access swift without keystone
+		// – not sure about the additional APIs
+		// https://docs.ceph.com/en/quincy/radosgw/config-ref/#confval-rgw_enable_apis
+		// see also https://docs.ceph.com/en/octopus/radosgw/config-ref/#swift-settings on disabling s3
+		// when using '/' as prefix
+
+		// Swift was enabled so far already by default, so perhaps better
+		// not change that if someone relies on it.
+
+		configOptions["rgw_enable_apis"] = "s3website, swift, swift_auth, admin, sts, iam, notifications"
+	}
+
 	for flag, val := range configOptions {
+		logger.Debugf("Set %s to %q on %q", flag, val, who)
 		err := monStore.Set(who, flag, val)
 		if err != nil {
+			logger.Debugf("failed to setup config option: %s to %q on %q: %s", flag, val, who, err)
 			return errors.Wrapf(err, "failed to set %q to %q on %q", flag, val, who)
 		}
 	}
 
 	return nil
 }
 
+func configureKeystoneAuthentication(rgwConfig *rgwConfig, configOptions map[string]string) (map[string]string, error) {
+
+	if keystone := rgwConfig.Auth.Keystone; keystone != nil {
```

Comment:
Golang encourages "happy path on the left" or fail-fast methods. In our experience, this makes for code that is much easier to read and less bug-prone. Please adjust to follow this pattern:

```go
keystone = rgwConfig.Auth.Keystone
if keystone == nil {
    logger.Debug("Authentication with keystone is disabled")
    return configOptions, nil
}

// everything that is inside the if block now goes directly here
```